### PR TITLE
Add a lockable context for improved memory safety.

### PR DESCRIPTION
**Pull request by connagh-opteran**
there are now functions for contexts that mirror the mutex functionality that they represent. these are: 
`lockContext`
`tryLockContext` 
`unlockContext`

The underlying implementation just uses a `mutex` to allow a thread to lock the context to its thread while performing GL commands.

### CODE REVIEW

File: morph/Visual.h [hunk 1]
```diff
@@ -255,6 +258,30 @@ namespace morph {
...
+        void lockContext() { 
+            context_mutex.lock(); 
+            setContext();
+        }
+        
+        //! Attempt to lock the context mutext returns if locking was successful. 
+        //! If success sets the context.
+        bool tryLockContext() { 
+          if(context_mutex.try_lock()){
+            set_context();
```
Comment (sebjameswml) [inline morph/Visual.h:271]: Shouldn't this be setContext()?

File: morph/Visual.h [hunk 2]
```diff
@@ -255,6 +258,30 @@ namespace morph {
         // A callback friendly wrapper
         static void set_context (morph::Visual<glver>* _v) { _v->setContext(); };
 
+        //! Lock the context mutext to prevent locking across multiple. Then sets the context.
+        void lockContext() { 
+            context_mutex.lock();
```
Comment (sebjameswml) [inline morph/Visual.h:263]: this->context_mutex and this->setContext

File: morph/Visual.h [hunk 3]
```diff
@@ -255,6 +258,30 @@ namespace morph {
         // A callback friendly wrapper
         static void set_context (morph::Visual<glver>* _v) { _v->setContext(); };
 
+        //! Lock the context mutext to prevent locking across multiple. Then sets the context.
+        void lockContext() {
```
Comment (sebjameswml) [inline morph/Visual.h:262]: { on newline after a function

File: morph/Visual.h [hunk 4]
```diff
@@ -255,6 +258,30 @@ namespace morph {
...
+        //! Lock the context mutext to prevent locking across multiple. Then sets the context.
+        void lockContext() { 
+            context_mutex.lock(); 
+            setContext();
+        }
+        
+        //! Attempt to lock the context mutext returns if locking was successful. 
+        //! If success sets the context.
+        bool tryLockContext() { 
+          if(context_mutex.try_lock()){
```
Comment (sebjameswml) [inline morph/Visual.h:270]: 4 spaces indentation, { on newline after function (but not after if). this-> specifiers to indicate class membership of attributes/functions

File: morph/Visual.h [hunk 5]
```diff
@@ -255,6 +258,30 @@ namespace morph {
...
+            set_context();
+            return true;
+          } else {
+            return false;
+          }
+
```
Comment (sebjameswml) [inline morph/Visual.h:276]: cull that extraneous blank line :)

**Review by sebjameswml — CHANGES_REQUESTED** [review]
Minor stuff to fix (4 spaces indentation, 'morphologica' style) plus I don't know why you call set_context()? Does that compile? Visual::set_context takes a Visual* argument and is intended only for inclusion in callbacks.

File: morph/Visual.h (hunk 1 above)
Comment (connagh-opteran) [reply morph/Visual.h]: yeah not sure how this got through but it did compile. updated to match the call.

File: morph/Visual.h (hunk 2 above)
Comment (connagh-opteran) [reply morph/Visual.h]: updated

File: morph/Visual.h (hunk 3 above)
Comment (connagh-opteran) [reply morph/Visual.h]: changed

File: morph/Visual.h (hunk 4 above)
Comment (connagh-opteran) [reply morph/Visual.h]: Changed all the above.

File: morph/Visual.h (hunk 5 above)
Comment (connagh-opteran) [reply morph/Visual.h]: done

**Review by sebjameswml — APPROVED** [review]
Looks good now. I'll merge this once the CI passes

**Comment by sebjameswml** [comment]
I'm getting "Merge attempt failed" "Merge already in progress" which I think is some sort of Github UI bug. Otherwise, this would already have been merged.

**Pull request merged**